### PR TITLE
feat: Implement screencast

### PR DESCRIPTION
**Pull request by jadaradix**
This PR implements screencasting.

We have 'battle tested' this code on OS X and inside production containers.

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -983,6 +987,21 @@ Emitted when a request finishes successfully.
 
 Emitted when a [response] is received.
 
+#### event: 'screencastframe'
+- <[object]>
```
Comment (vsemozhetbyt) [inline docs/api.md]: Non-primitive types seem uppercased:
```suggestion
- <[Object]>
```

File: docs/api.md [hunk 2]
```diff
@@ -983,6 +987,21 @@ Emitted when a request finishes successfully.
...
+    - `data` <[string]> Base64 encoded frame data.
+    - `metadata` <[object]>
```
Comment (vsemozhetbyt) [inline docs/api.md]:
```suggestion
    - `metadata` <[Object]>
```

File: docs/api.md [hunk 3]
```diff
@@ -1559,6 +1578,10 @@ Shortcut for [page.mainFrame().executionContext().queryObjects(prototypeHandle)]
     - `networkidle2` - consider navigation to be finished when there are no more than 2 network connections for at least `500` ms.
 - returns: <[Promise]<[Response]>> Promise which resolves to the main resource response. In case of multiple redirects, the navigation will resolve with the response of the last redirect.
 
+#### page.screencastFrameAck(sessionId)
+- `sessionId` <[number]> The ID of the frame that is being acknowledged. Provided in the `screencastframe` event.
```
Comment (vsemozhetbyt) [inline docs/api.md:1582]:
```suggestion
- `sessionId` <[number]> The ID of the frame that is being acknowledged. Provided in the `'screencastframe'` event.
```

File: docs/api.md [hunk 4]
```diff
@@ -1724,6 +1747,18 @@ puppeteer.launch().then(async browser => {
 
 In the case of multiple pages in a single browser, each page can have its own viewport size.
 
+#### page.startScreencast(options)
+- `options` <[Object]>
+  - `format` <"png" | "jpeg"> Specify screencast frame format, could be either `jpeg` or `png`. Defaults to 'png'.
```
Comment (vsemozhetbyt) [inline docs/api.md]:
```suggestion
  - `format` <"png" | "jpeg"> Specify screencast frame format, could be either `'jpeg'` or `'png'`. Defaults to `'png'`.
```

File: docs/api.md [hunk 5]
```diff
@@ -1724,6 +1747,18 @@ puppeteer.launch().then(async browser => {
...
+  - `quality` <[number]> The quality of the image, between 0-100. Not applicable to `png` images.
```
Comment (vsemozhetbyt) [inline docs/api.md:1753]:
```suggestion
  - `quality` <[number]> The quality of the image, between 0-100. Not applicable to `'png'` images.
```

**Review by vsemozhetbyt — COMMENTED** [review]
Thank you. I've just spotted some tiny consistency nits in doc format.

File: docs/api.md (hunk 3 above)
Comment (jadaradix) [reply docs/api.md:1582]: I don't think we need extra '' here, e.g.

"Sends a `keydown`, `keypress`/`input`, and `keyup` event for each character in the text."

File: docs/api.md (hunk 5 above)
Comment (jadaradix) [reply docs/api.md:1753]: `png` is consistent with the docs @ `page.screenshot([options])` :-)

**Comment by aslushnikov** [comment]
Hey @jadaradix, thanks for the PR.

I still think that exposing mpg stream from chromium would be a way nicer approach - less data to transfer over the wire, and much more control over the video format.

Can you publish this as a third-party package atop of Puppeteer?

**Comment by jadaradix** [comment]
> Hey @jadaradix, thanks for the PR.
> 
> **I still think that exposing mpg stream from chromium would be a way nicer approach** - less data to transfer over the wire, and much more control over the video format.
> 
> Can you publish this as a third-party package atop of Puppeteer?

Hi @aslushnikov! This is something we'd like to do here. Can you point me in the right direction re: mpg stream? If I have an API reference I will implement it in Puppeteer and we can decide which feature/PR to merge.

I added a test case for screen casting too @ https://github.com/GoogleChrome/puppeteer/blob/223244717b9cfa38f1adf6fe0f36d891d00fe66c/test/screencast.spec.js

**Comment by aslushnikov** [comment]
@jadaradix the links that should help start with Chrome DevTools work:
- [gettings started with DevTools protocol](https://github.com/aslushnikov/getting-started-with-cdp)
- [contributing to DevTools protoco](https://docs.google.com/document/d/1c-COD2kaK__5iMM5SEx-PzNA7HFmgttcYfOHHX0HaOM/edit#heading=h.e6mz7k1mw34a)
- [devtools protocol viewer](https://vanilla.aslushnikov.com/)
- [chromium code search](https://cs.chromium.org)

Closing this PR: the screencast we want is the MPG stream from chromium, so all the work should be done on the browser side.

**Comment by 0xcaff** [comment]
Did anyone ever start the work for exposing the MPG stream from chromium?

It seems that someone has created a [`playwright-video`](https://github.com/qawolf/playwright-video) library implementing screencast capture as a third party package for playwright. I've created a fork with support for puppeteer (https://github.com/0xcaff/playwright-video/tree/feature/puppeteer) and am using that for now.

**Pull request closed without merging**